L1Trigger/L1THGCal/src/be_algorithms/HGCalClusteringImpl.cc

```diff
@@ -270,3 +276,162 @@ void HGCalClusteringImpl::clusterizeNN( const edm::PtrVector<l1t::HGCalTriggerCe
+                                        l1t::HGCalClusterBxCollection & clusters,
+                                        const HGCalTriggerGeometryBase & triggerGeometry
+    ){
+
+    bool isSeed[triggerCellsPtrs.size()];
+    std::vector<unsigned> seedPositions;
+    seedPositions.reserve( triggerCellsPtrs.size() );
+
+    /* search for cluster seeds */
+    int itc(0);
+    for( edm::PtrVector<l1t::HGCalTriggerCell>::const_iterator tc = triggerCellsPtrs.begin(); tc != triggerCellsPtrs.end(); ++tc,++itc ){
+        
+        double seedThreshold = ((*tc)->subdetId()==HGCHEB ? scintillatorSeedThreshold_ : siliconSeedThreshold_);
+
+        /* decide if is a seed, if yes store the position into of triggerCellsPtrs */
+        isSeed[itc]       = ( (*tc)->mipPt() > seedThreshold) ? true : false;
+        if( isSeed[itc] ) seedPositions.push_back( itc );
+
+        /* remove tc from the seed vector if is a NN of an other seed*/
+        for( auto pos : seedPositions ){
+            if( ( (*tc)->position() - triggerCellsPtrs[pos]->position() ).mag()  < dr_ ){
+                if( this->areTCneighbour_( (*tc)->detId(), triggerCellsPtrs[pos]->detId(), triggerGeometry ) )
+                {
+                    isSeed[itc] = false;
+                    seedPositions.pop_back();
+                }
+            }
+        } 
+
+    }
+    
+    /* clustering the TCs */
+    std::vector<l1t::HGCalCluster> clustersTmp;
+
+    // every seed generates a cluster
+    for( auto pos : seedPositions ) {
+        clustersTmp.emplace_back( triggerCellsPtrs[pos] );
+    }
+
+    /* add the tc to the clusters */
+    itc=0;
+    for( edm::PtrVector<l1t::HGCalTriggerCell>::const_iterator tc = triggerCellsPtrs.begin(); tc != triggerCellsPtrs.end(); ++tc,++itc ){
+      
+        /* get the correct threshold for the different part of the detector */ 
+        double threshold = ((*tc)->subdetId()==HGCHEB ? scintillatorTriggerCellThreshold_ : siliconTriggerCellThreshold_);
+        
+        /* continue if not passing the threshold */
+        if( (*tc)->mipPt() < threshold ) continue;
+        
+        /* searching for TC near the centre of the cluster  */
+        std::vector<unsigned> tcPertinentClusters; 
+        unsigned iclu(0);
+        
+        for ( const auto& clu : clustersTmp ) {
+            if( this->isPertinent(**tc, clu, dr_) ){
+                tcPertinentClusters.push_back( iclu );
+            }
+            ++iclu;
+        }   
+
+        if ( tcPertinentClusters.size() == 0 ) {
+            continue;
+        }
+        else if( tcPertinentClusters.size() == 1 ) {
+            clustersTmp.at( tcPertinentClusters.at(0) ).addConstituent( *tc );
+        }
+        else {
+
+            /* calculate the fractions */
+            double totMipt = 0;
+            for( auto clu : tcPertinentClusters ){
+                totMipt += clustersTmp.at( clu ).constituents()[0]->mipPt();
```

Comment:
`seedMipPt()` should give you the same thing as `constituents()[0]->mipPt()` and is easier to read.